src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/DynamicEvents/DynamicEvents.cs

```diff
@@ -0,0 +1,129 @@
+// TODO, AndrewAu, remove this condition when new TraceEvent is available.
+#if CUSTOM_TRACE_EVENT
+
+using Microsoft.Diagnostics.Tracing.Analysis.GC;
+using Microsoft.Diagnostics.Tracing.Parsers.GCDynamic;
+using System.Diagnostics;
+using System.Dynamic;
+
+namespace GC.Analysis.API.DynamicEvents
+{
+
+    public static class TraceGCExtensions
+    {
+        public static dynamic DynamicEvents(this TraceGC traceGC)
+        {
+            return new DynamicIndex(traceGC.dynamicEventIndex);
+        }
+    }
+
+    // TODO, AndrewAu, this could be compiled for better performance, right now we are doing linear lookups
+    // We can do validation so that the input make sense (i.e. no unexpected types, no duplicate event/field), etc
+    //
+    public class DynamicEventSchema
+    {
+        internal static List<DynamicEventSchema> DynamicEventSchemas = new List<DynamicEventSchema>();
+
+        public string DynamicEventName { get; set; }
+
+        public List<KeyValuePair<string, Type>> Fields { get; set; }
+
+        public static void Set(List<DynamicEventSchema> dynamicEventSchemas)
+        {
+            DynamicEventSchemas = dynamicEventSchemas;
+        }
+    }
+
+    internal class DynamicIndex : DynamicObject
+    {
+        private Dictionary<string, List<RawDynamic>> index;
+
+        public DynamicIndex(Dictionary<string, List<RawDynamic>> index)
+        {
+            this.index = index;
+        }
+
+        public override bool TryGetMember(GetMemberBinder binder, out object result)
+        {
+            if (DynamicEventSchema.DynamicEventSchemas.Any(s => string.Equals(s.DynamicEventName, binder.Name)))
+            {
+                string name = binder.Name;
+                if (index.TryGetValue(name, out var indexValue))
+                {
+                    // TODO, andrewau, handle the case where we have multiple dynamic event of the same name in a single GC.
+                    //
+                    // That should be an array of these dynamic events
+                    // Somewhere the schema should tell us if this is expected or not so not to generate wrong stuff just because the data is corrupted.
+                    // 
+                    result = new DynamicEvent(indexValue[0], DynamicEventSchema.DynamicEventSchemas.Single(s => string.Equals(s.DynamicEventName, binder.Name)).Fields);
```

Comment:
Single will except if a unique or no item is found. Do we really want to except here without context? I would do a `SingleOrDefault` or try-catch the error and then contextualize it with other details vs. a generic call stack.

Reply:
Yes, I am still trying to figure out the pattern here, something along the line of having the schema to provide some constraints, so this can do the proper handling.

Something like:

```c#
enum EventConstraint
{
  ExactlyOne,
  AtMostOne,
  OneOrMore,
  ZeroOrMore
}
```

Reply:
The "More" option is what I really need to think more about - imagine we emit two dynamic events of the same type before and after a GC. We probably would like to be able to get an array of them. Right now we can't.